components/eamxx/src/share/grid/remap/vertical_remapper.cpp

```diff
@@ -0,0 +1,286 @@
+  auto tgt = FieldLayout::invalid();
+  const bool midpoints = src_layout.has_tag(LEV);
+  const int vec_dim = src_layout.is_vector_layout() ? src_layout.dim(CMP) : -1;
+  switch (lt) {
+    case LayoutType::Scalar2D:
+      tgt = m_tgt_grid->get_2d_scalar_layout();
+      break;
+    case LayoutType::Vector2D:
+      tgt = m_tgt_grid->get_2d_vector_layout(CMP,vec_dim);
+      break;
+    case LayoutType::Scalar3D:
+      tgt = m_tgt_grid->get_3d_scalar_layout(midpoints);
+      break;
+    case LayoutType::Vector3D:
+      tgt = m_tgt_grid->get_3d_vector_layout(midpoints,CMP,vec_dim);
+      break;
+    default:
+      EKAT_ERROR_MSG ("Layout not supported by VerticalRemapper: " + e2str(lt) + "\n");
+  }
+  return tgt;
+}
+
+void VerticalRemapper::
+set_pressure_levels(const std::string& map_file) {
+  // TODO: I have concerns here with packsize.  I am using SCREAM_PACK_SIZE, which appears to work
+  //       but maybe it is possible for different variables to have different packsizes, and if so
+  //       then we can't hardcode the packsize for the target_levels.  Can we define as REAL and
+  //       still get_view using a different packsize?  Or some other strategy?
+  scorpio::register_file(map_file,scorpio::FileMode::Read);
+  m_num_remap_levs = scorpio::get_dimlen_c2f(map_file.c_str(),"nlevs");
+
+  auto npacks = ekat::PackInfo<SCREAM_PACK_SIZE>::num_packs(m_num_remap_levs);
+  m_remap_pres_view = view_1d<Pack>("",npacks);
+  auto remap_pres_scal = ekat::scalarize(m_remap_pres_view);
+
+  std::vector<scorpio::offset_t> dofs_offsets(m_num_remap_levs);
+  std::iota(dofs_offsets.begin(),dofs_offsets.end(),0);
+  const std::string idx_decomp_tag = "vertical_remapper::" + std::to_string(m_num_remap_levs);
+  scorpio::get_variable(map_file, "p_levs", "p_levs", {"nlevs"}, "real", idx_decomp_tag);
+  scorpio::set_dof(map_file,"p_levs",m_num_remap_levs,dofs_offsets.data());
+  scorpio::set_decomp(map_file);
+  scorpio::grid_read_data_array(map_file,"p_levs",-1,remap_pres_scal.data(),remap_pres_scal.size());
+  scorpio::eam_pio_closefile(map_file);
+}
+
+void VerticalRemapper::
+register_vertical_source_field(const Field& src, const std::string& mode)
+{
+  using namespace ShortFieldTagsNames;
+  EKAT_REQUIRE_MSG(mode=="mid" || mode=="int","Error: VerticalRemapper::register_vertical_source_field,"
+    "mode arg must be 'mid' or 'int'\n");
+
+  auto src_fid = src.get_header().get_identifier();
+  if (mode=="mid") {
+    auto layout = src_fid.get_layout();
+    auto name   = src_fid.name();
+    EKAT_REQUIRE_MSG(ekat::contains(std::vector<FieldTag>{LEV},layout.tags().back()),
+      "Error::VerticalRemapper::register_vertical_source_field,\n"
+      "mode = 'mid' expects a layour ending with LEV tag.\n"
+      " - field name  : " + name + "\n"
+      " - field layout: " + to_string(layout) + "\n");
+    EKAT_REQUIRE_MSG(src.is_allocated(), "Error! LEV source field is not yet allocated.\n");
+    src_mid = src;
```

Comment:
To follow other member vars, you could prepend `m_` to member var names. It helps to distinguish immediately if a var is local or a member, when looking at a code snippet.